docs/design/2019-11-14-tidb-builtin-diagnostics-zh_CN.md

```diff
@@ -0,0 +1,715 @@
+    - wsec/s: 每秒写取的扇区数
+    - avgrq-sz: 平均请求扇区的大小
+    - avgqu-sz: 是平均请求队列的长度
+    - await: 每一个IO请求的处理的平均时间（单位是微秒毫秒）
+    - svctm: 表示平均每次设备I/O操作的服务时间（以毫秒为单位）
+    - %util: 在统计时间内所有处理IO时间，除以总共统计时间
+- 网络 IO
+    - IFACE：LAN接口
+    - rxpck/s：每秒钟接收的数据包
+    - txpck/s：每秒钟发送的数据包
+    - rxbyt/s：每秒钟接收的字节数
+    - txbyt/s：每秒钟发送的字节数
+    - rxcmp/s：每秒钟接收的压缩数据包
+    - txcmp/s：每秒钟发送的压缩数据包
+    - rxmcst/s：每秒钟接收的多播数据包
+- 常用的系统配置：sysctl -a
+
+#### 节点配置信息
+
+所有节点都包含当前节点的生效配置，不需要额外的步骤既可拿到配置信息。
+
+#### 节点日志信息
+
+TiDB/TiKV/PD 产生的日志都保存在各自的节点上，并且 TiDB 集群部署过程中没有部署额外的日志收集组件，所以在日志检索中有以下问题：
+
+- 日志分布在各个节点，需要单独登陆到每一个节点使用关键字进行搜索
+- 日志文件会每天 rotate，所以在单个节点也需要对多个日志文件进行搜索
+- 没有简单的方式对多个节点的日志按照时间排序整合到同一个文件
+
+本提案提供以下两种思路来解决以上问题：
+
+- 引入第三方日志收集组件对所有节点的日志进行收集
+    - 优势：统一的日志管理，日志可以长时间保存，并易于检索，并且多个组件的日志可以按照时间排序归并
+    - 劣势：增加集群运维难度，第三方组件不容易与 TiDB 内部 SQL 集成；日志收集工具会收集全量日志，收集过程占用各个系统资源（磁盘 IO、网络 IO）
+- 各个节点提供日志服务，TiDB 通过各个节点的接口将谓词下推到日志检索接口，直接对各个节点返回的日志进行归并
+    - 优势：不引入三方组件，谓词下推后只返回过滤后的日志，能轻易的与 TiDB SQL 进行集成，并能复用 SQL 引擎的过滤、聚合等
+    - 劣势：如果节点日志删除后，不能检索到对应日志
+
+根据以上的优劣势分析，本提案使用第二种方案，即各个节点提供日志搜索接口，TiDB 将日志搜索的 SQL 中谓词下推到各个节点，日志搜索接口的语义为：搜索本地日志文件，并使用谓词进行过滤，匹配的结果返回。
+
+- `start_time`: 日志检索的开始时间（unix 时间戳，单位毫秒），如果没有该谓词，则默认为 0。
+- `end_time`: 日志检索的开始时间（unix 时间戳，单位毫秒），如果没有该谓词，则默认为 `int64::MAX`。
+- `pattern`: 如 SELECT * FROM tidb_cluster_log WHERE pattern LIKE "%gc%" 中的 %gc% 即为过滤的关键字
+- `level`: 日志等级，可以选为 DEBUG/INFO/WARN/WARNING/TRACE/CRITICAL/ERROR
+- `limit`: 返回日志的条数，如果没有指定，则限制为 64k 条，防止日质量太大占用大量网络
+
+#### 节点性能采样数据
+
+当前 TiDB 集群中，发现有性能瓶颈时，需要快速定位问题。火焰图 （Flame Graph）是由 Brendan Gregg 发明的，与其他的 trace 和 profiling 方法不同的是，Flame Graph 以一个全局的视野来看待时间分布，它从底部往顶部，列出所有可能的调用栈。其他的呈现方法，一般只能列出单一的调用栈或者非层次化的时间分布。
+
+目前 TiKV 和 TiDB 获取火焰图的方式不同，并且都需要依赖外部工具。
+
+- TiKV 获取火焰图
+
+    ```
+    perf record -F 99 -p proc_pid -g -- sleep 60
+    perf script > out.perf
+    /opt/FlameGraph/stackcollapse-perf.pl out.perf > out.folded
+    /opt/FlameGraph/flamegraph.pl out.folded > cpu.svg
+    ```
+
+- TiDB 获取火焰图
+
+    ```
+    curl http://127.0.0.1:10080/debug/pprof/profile > cpu.pprof
+    go tool pprof -svg cpu.svn cpu.pprof
+    ```
+
+目前存在的两个主要问题：
+
+- 生产环境中不一定包含对应的外部工具（perf/flamegraph.pl/go）
+- TiKV 和 TiDB 没有统一的方式
+
+为了解决以上两个问题，本提案将获取火焰图的方法内置到 TiDB 中，统一使用 SQL 触发采样并将采样数据转换为火焰图作为查询结果显示，一方面降低对外部工具的依赖，同时也极大的提升效率。各个节点实现采样数据采集功能并提供采样接口，对上层输出指定格式的采样数据。暂定输出为 `[pprof](github.com/google/pprof)` 定义的 ProtoBuf 格式。
+
+采样数据获取方式：
+
+- TiDB/PD: 使用 Golang Runtime 内置的采样数据获取接口
+- TiKV: 使用 `[pprof-rs](github.com/tikv/pprof-rs)` 库采集采样数据
+
+#### 节点监控信息
+
+监控信息主要是各个组件内部定义的监控指标。目前 TiDB/TiKV/PD 都会提供 `/metrics` HTTP API，然后通过部署的 Prometheus 组件定时（默认配置 15s）的拉取集群各个节点的监控指标。并且部署了 Grafana 组件用于从 Prometheus 拉取监控数据，进行可视化展示。
+
+监控信息不同于实时获取的系统信息，监控数据是一个时序数据。包含各个节点在各个时间点的数据，对于排查问题和诊断问题有非常重要的用途，所以监控信息的保存和查询对于本提案实现 TiDB 内置 SQL 诊断非常重要。为了能够在 TiDB 内使用 SQL 查询监控数据，目前有以下备选方案：
+
+- 使用 Prometheus client 和 PromQL 查询 Prometheus server 的数据
+    - 优势：有现成解决方案，只需要将 Prometheus server 的地址注册到 TiDB 即可，实现简单
+    - 劣势：增强了 TiDB 对 Prometheus 的依赖，为后续完全移除 Prometheus 增加了困难
```

Comment:
I am curious about how to integrate the alert and graphing solutions from community, like alertmanager and grafana after we remove Prometheus.